alws/crud.py

```diff
@@ -782,3 +788,228 @@ async def commit_release(db: Session, release_id: int) -> (models.Release, str):
         selectinload(models.Release.platform)
     ))
     return release_res.scalars().first(), message
+
+
+async def get_repositories(db: Session, repository_id: int = None
+                           ) -> typing.List[models.Repository]:
+    repo_q = select(models.Repository)
+    if repository_id:
+        repo_q = repo_q.where(models.Repository.id == repository_id)
+    result = await db.execute(repo_q)
+    return result.scalars().all()
+
+
+async def create_repositories(
+        db: Session,
+        payload: typing.List[repository_schema.RepositoryCreate]
+) -> typing.List[models.Repository]:
+    # We need to update existing repositories instead of trying to create
+    # new ones if they have the same parameters
+    query_list = [
+        sqlalchemy.and_(
+            models.Repository.name == item.name,
+            models.Repository.arch == item.arch,
+            models.Repository.type == item.type
+        )
+        for item in payload
+    ]
+    query = sqlalchemy.or_(*query_list).with_for_update()
+    repos_mapping = {}
+    async with db.begin():
+        repos_result = await db.execute(query)
+        for repo in repos_result.scalars().all():
+            repo_key = f'{repo.name}-{repo.arch}-{repo.debug}'
+            repos_mapping[repo_key] = repo
+
+        for repo_item in payload:
+            repo_item_dict = repo_item.dict()
+            repo_key = f'{repo_item.name}-{repo_item.arch}-{repo_item.debug}'
+            if repo_key not in repos_mapping:
+                repos_mapping[repo_key] = models.Repository(**repo_item_dict)
+            else:
+                repo = repos_mapping[repo_key]
+                for field, value in repo_item_dict.items():
+                    setattr(repo, field, value)
+
+        db.add_all(repos_mapping.values())
+        await db.commit()
+
+    for repo in repos_mapping.values():
+        await db.refresh(repo)
+
+    return list(repos_mapping.values())
+
+
+async def create_repository(
+        db: Session, payload: repository_schema.RepositoryCreate
+) -> models.Repository:
+    query = select(models.Repository).where(
+        models.Repository.name == payload.name,
+        models.Repository.arch == payload.arch,
+        models.Repository.type == payload.type,
+        models.Repository.debug == payload.debug,
+    )
+    async with db.begin():
+        result = await db.execute(query)
+        if result.scalars().first():
+            raise ValueError('Repository already exists')
+
+        repository = models.Repository(**payload.dict())
+        db.add(repository)
+        await db.commit()
+    await db.refresh(repository)
+    return repository
+
+
+async def update_repository(
+        db: Session, repository_id: int,
+        payload: repository_schema.RepositoryUpdate
+) -> models.Repository:
+    async with db.begin():
+        result = await db.execute(select(models.Repository).where(
+            models.Repository.id == repository_id))
```

Comment:
Why do you not use `get(repository_id`?

Reply:
https://stackoverflow.com/a/51997031 in this particular case you're right, but it's not always a correct way. I'll fix it